### ISSUE TITLE
[ 2.9 ]  float64 min

### ISSUE THREAD

**Issue by wyNomad** [flagos-ai/Torch-FL#274]
## Issue Type
- [x] Bug Report
- [ ] Documentation

## AI Agent Information
- **Agent**: Claude Code CLI
- **Model**: Claude Fable 5
- **Session Context**: While wiring the Kunlunxin P800 CI line on the `2.9` stable branch of `flagos-ai/Torch-FL`, the runtime smoke harness (`tests/manual/kunlun_runtime_smoke.py`) stochastically aborted during the device-allocation check. The abort originates inside the vendor XDNN library, not in our adapter code. This issue reports the XDNN-internal fp64 `min` capability probe that causes it.

## Summary

`torch.empty((8, 8), device="flagos:1")` **stochastically** aborts the whole Python process from C++ via `exit(1)`. The stack lands inside the vendor XDNN library at `min.cpp:41` with `scalar type of self : kfloat64 is unsupported`. Python `try/except` cannot catch it. The abort is non-deterministic: the **exact same binary, same env, same code** passes on one CI run and aborts on the next.

The root cause is a **self-contradictory capability probe inside XDNN**: during the `torch.empty` allocation path, XDNN internally constructs a `float64` tensor and runs `min` on it to probe whether the device supports fp64 — but XDNN's own `min` does not support fp64, so the probe kills the process. **The user never requested float64** (`torch.empty((8,8))` defaults to `float32`); the fp64 tensor is fabricated by XDNN's internal probe.

## Error

```
XDNN_PYTORCH ... min.cpp:41
scalar type of self : kfloat64 is unsupported
```
- C++-level `exit(1)` — Python `try/except` cannot intercept it.
- The full smoke harness (9 checks) dies at the `allocation` check, which is the first check that touches device memory:
  ```python
  check("allocation", lambda: torch.empty((8, 8), device=DEVICE))  # DEVICE = "flagos:1"
  ```

## Critical Clarification: The User Did NOT Request float64

This is the most important point and the easiest to misread. The error says `kfloat64`, which invites the conclusion that "the test called a float64 min." **It did not.**

- `torch.empty((8, 8))` with no `dtype=` argument uses PyTorch's default dtype, which is **`float32`**.
- The entire `kunlun_runtime_smoke.py` file contains **zero** `float64`/`torch.double` references and **zero** explicit `min` calls.
- The `float64` in the error is **fabricated by XDNN itself** — XDNN constructs a fp64 tensor internally during the `empty` allocation path as a capability probe, then calls `min` on it.

In one sentence: **XDNN uses an operator it does not support (`min` on `fp64`) to probe a capability it does not support (fp64), and the probe kills our process.**

## Evidence the Probe Is Op-Path-Bound, Not Dtype-Bound

A controlled in-process warm-up (run immediately before the smoke, in the same process) prints each step:

```python
torch_fl.flagos.init();            print("init done")            # PASS
torch.randn(1024,1024, device="flagos:0"); print("device0 randn done")  # PASS
torch.randn(1024,1024, device="flagos:1"); print("device1 randn done")  # PASS
# then runpy.run_path("tests/manual/kunlun_runtime_smoke.py", run_name="__main__")
#   → smoke's allocation check: torch.empty((8,8), device="flagos:1") → exit(1) at min.cpp:41
```

`randn` and `empty` are **both `float32`** (neither specifies a dtype). `randn` on `flagos:1` passes; `empty` on `flagos:1` aborts. **Same dtype, same device, same process — different op.** This proves the probe is bound to the `torch.empty` allocation path, **not** to the requested dtype. Therefore "just use float32 instead" is a no-op: `empty` is already float32; the fp64 is XDNN-internal.

## Stochastic Nature (the same code passes and fails)

The probe triggers non-deterministically. On upstream PR #273, the **same commit `95ee2b3`** (identical base `07c5c96` + identical head; merge SHA differs only by GitHub committer timestamp, **tree is byte-identical**) produced opposite results across two runs:

| CI run | Time (UTC) | Runner pod | Test step | Result |
|---|---|---|---|---|
| #635 | 08:20 | `…6d5bl-runner-jq5kn` | 102s, ran all 6 items | **6/6 green** |
| #641 | 09:44 | `…6d5bl-runner-lw9gq` | 35s, early abort | **red** — abort at smoke `min.cpp:41` |

Both runs had `XPU_CUPTI_ENABLE_DEVICE` delivered to the smoke step (corroborated by the absence of the `No devices are setting for profiling` CUPTI warning in step [5], which appears in steps [1][2][3][4][6] when the env is absent). The abort in #641 happened **with env + init both present** — confirming the probe is stochastic and the env+init mitigation is **probabilistic, not deterministic**.

## Three CI Controlled Experiments (how we isolated the probe)

We did not modify the smoke file. We changed only CI config, one variable at a time:

| commit | `XPU_CUPTI_ENABLE_DEVICE` (global) | in-process `init()` before smoke | smoke [5/6] | dispatch [6/6] | conclusion |
|---|---|---|---|---|---|
| `a089229` | set `0..7` | no | **abort** (empty) | error 17 | env alone is not enough |
| `2bbdbfa` | set `0..7` | yes (is_available+init+randn+runpy) | **PASS** | error 17 | env+init together pass smoke |
| `a0e0d5d` | unset (removed) | yes | **abort** (empty) | not reached | init alone is not enough |

Two findings:
1. **Only `empty` triggers the probe; `randn` does not.** (The dispatch subprocesses' first op is `randn`, so they do not need the env.)
2. **Suppressing the `empty` probe requires `env + init` together** — env alone aborts, init alone aborts, both together pass (probabilistically).

## Workaround We Applied (and its limitation)

We split the env's scope (commit `95ee2b3`):
- **Global**: remove `XPU_CUPTI_ENABLE_DEVICE` from `set_env_kunlun.sh` + `GITHUB_ENV` → dispatch subprocesses (first op `randn`, no probe) inherit no env → no CUPTI `error 17`.
- **Smoke-local**: `export XPU_CUPTI_ENABLE_DEVICE=0,1,2,3,4,5,6,7` only inside the smoke step's bash subshell + in-process `init()` + `randn` warm-up on device 0 and 1 + `runpy.run_path(smoke, run_name="__main__")` to run the unmodified smoke source in the already-initialized process.

This **unblocked the smoke × error17 deadlock** (a real technical result). **But it is non-deterministic** — run #641 proves the probe still fires with env+init present. The mitigation is probabilistic suppression, not a fix.

The env's dual nature is the source of the deadlock:
- Without env → smoke `empty` aborts (XDNN fp64 `min` probe).
- With env `0..7` → smoke passes (probe suppressed) **but** dispatch hits CUPTI `error 17` on every subprocess (`cuptiActivityEnable` → `etiEventSamplingSetMode(ctx, true)` fails: multi-device CUPTI event sampling unsupported + container privilege limits).

## Why This Cannot Be Fixed on Our Side

1. **The probe is inside the XDNN binary library.** We cannot modify the vendor library source. It injects the fp64 `min` probe into the `empty` allocation path; that is the library's behavior.
2. **The trigger is stochastic**, governed by XDNN's internal state machine. `env + init` only tunes external preconditions; it cannot deterministically control when XDNN decides to probe.
3. **XDNN's `min` does not support fp64, yet XDNN uses fp64 `min` as the probe** — this is a self-contradictory bug inside the vendor library. Only the vendor can resolve it: make `min` support fp64, or stop using fp64 `min` (or any `min`) as the capability probe.

Three checks confirm the abort is **inside the vendor library, not in torch_fl**:
1. **No `aten::min` call in our csrc** — searching `csrc/` (excluding `generated/`) returns only C++ STL `std::min` on integers (`csrc/profiler/cupti_device_tracer.cc`, `csrc/runtime/hooks.h`, `csrc/aten/backends/ascend/split.cc`). No `at::min`/`torch::min`/`aten::min`.
2. **The profiler tracer is pure C++ STL** — `csrc/profiler/cupti_device_tracer.cc` has no tensor call.
3. **`XPU_CUPTI_ENABLE_DEVICE` has zero references in csrc/torch_fl** — it is read by the vendor XPU-RT/XDNN library, not by our code.

## Questions for the Vendor

1. Does XDNN run an **fp64 capability probe** during the `torch.empty` (device memory allocation) path? Why?
2. Why does the probe use **fp64 `min`** when XDNN's own `min` does not support fp64? This is self-contradictory.
3. Can this be fixed at the library level — either (a) make `min` support fp64, or (b) change the probe to not use fp64 / not use `min`?
4. Is there an **environment variable to disable the fp64 capability probe** (analogous to how `XPU_CUPTI_ENABLE_DEVICE` controls profiler binding)? If so, a single env var would deterministically bypass the abort.
5. Why is the probe **stochastic**? What internal state decides whether the probe fires, given that env + init are both present?
6. (Doc contradiction) `docs/vendors/kunlun/installation.md:65` documents `XPU_CUPTI_ENABLE_DEVICE` as **optional** and states an unset value "does not prevent ordinary allocation or operator execution." On driver `5.0.21.47` / XPU-RT `5.37.1` this is no longer true — an unset value stochastically aborts first device-context init via this fp64 `min` probe. The doc should be corrected, or the probe removed/fixed so the doc statement holds.

## Environment
<details>
<summary>Click to expand environment details</summary>

- **Platform**: Kunlunxin (P800, CUDA boxing path, `ACCELERATOR=kunlun`)
- **Python**: `/flagos/bin/python` 3.10.20
- **PyTorch**: `2.9.0+cu129` (vendor), `torch.version.cuda == "12.9"`
- **torch_fl**: built from source on `2.9` stable branch
- **Hardware**: 8× P800 OAM, driver `5.0.21.47`, XPU-RT `5.37.1`
- **Image**: `harbor.baai.ac.cn/flagos-runtime/flagos-runtime-kunlunxin-xre5.37.1:2.1.2`
- **Runner label**: `p800-8g-cicd-pytorch` (self-hosted, 8 devices per node, shared physical node `6d5bl` across pods)
- **Container**: `--device /dev/xpu0..7` + `--device /dev/xpuctrl` + `--shm-size=64g --privileged --security-opt seccomp=unconfined`
</details>

## Reproduction

```bash
# Minimal repro — torch.empty on a flagos device stochastically aborts:
export ACCELERATOR=kunlun XPU_ROOT=/usr/local/xpu XCUDART_ROOT=/usr/local/xcudart \
       XPU_ENABLE_PROFILER_TRACING=1
/flagos/bin/python -c "
import torch, torch_fl
torch_fl.flagos.init()
# randn does NOT trigger the probe:
torch.randn(1024, 1024, device='flagos:0'); print('randn ok')
torch.randn(1024, 1024, device='flagos:1'); print('randn device1 ok')
# empty (same float32) stochastically triggers the XDNN fp64 min probe:
torch.empty((8, 8), device='flagos:1'); print('empty ok')
"
# expected on abort: XDNN_PYTORCH min.cpp:41  scalar type of self : kfloat64 is unsupported  → exit(1)
# (non-deterministic: may pass; re-run / vary device state to reproduce)

# With env (probabilistic suppression, NOT a deterministic fix):
export XPU_CUPTI_ENABLE_DEVICE=0,1,2,3,4,5,6,7
/flagos/bin/python -c "
import torch, torch_fl
torch_fl.flagos.init()
torch.empty((8, 8), device='flagos:1'); print('empty ok')
"
# usually passes, but run #641 proves it can still abort with env+init present
```

## Related Code Locations
- `tests/manual/kunlun_runtime_smoke.py:39` — the `allocation` check (`torch.empty((8,8), device=flagos:1)`) that triggers the probe.
- `torch_fl/configs/backends_cuda.conf` — `min = cuda` route (the route the implicit `min` takes to XDNN).
- `.github/configs/kunlunxin.yml` — smoke command (local `export XPU_CUPTI_ENABLE_DEVICE=0..7` + in-process init + runpy).
- `.github/scripts/set_env_kunlun.sh` — global env is **not** set here (removed to avoid CUPTI error 17 in dispatch).
- `docs/vendors/kunlun/installation.md:65` — the `XPU_CUPTI_ENABLE_DEVICE` "optional" claim
